### PR TITLE
Testing-Tool with simulation Operators

### PR DESCRIPTION
**Pull request by katja-seb**

### CODE REVIEW

**Comment by hirmerpl** [comment]
In ubuntu, the following exception is thrown: 
java.awt.HeadlessException
	java.awt.GraphicsEnvironment.checkHeadless(GraphicsEnvironment.java:204)
	java.awt.Window.<init>(Window.java:536)
	java.awt.Frame.<init>(Frame.java:420)
	java.awt.Frame.<init>(Frame.java:385)
	javax.swing.JFrame.<init>(JFrame.java:189)
	org.citopt.connde.service.testing.GraphPlotter.<init>(GraphPlotter.java:33)
	sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:142)
	org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:89)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateBean(AbstractAutowireCapableBeanFactory.java:1102)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1054)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:512)
	org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482)
	org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306)
	org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230)
	org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302)
	org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197)
	org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:754)
	org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:866)
	org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:542)
	org.springframework.web.servlet.FrameworkServlet.configureAndRefreshWebApplicationContext(FrameworkServlet.java:668)
	org.springframework.web.servlet.FrameworkServlet.initWebApplicationContext(FrameworkServlet.java:540)
	org.springframework.web.servlet.FrameworkServlet.initServletBean(FrameworkServlet.java:494)
	org.springframework.web.servlet.HttpServletBean.init(HttpServletBean.java:136)
	javax.servlet.GenericServlet.init(GenericServlet.java:158)

File: src/main/java/org/citopt/connde/service/testing/TestEngine.java [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+     * Opens the Test-Report
+     *
+     * @param testId ID of the specific test
+     */
+    public ResponseEntity<String> openPDF(String testId) {
+
+        ResponseEntity<String> response;
+        try {
+            TestDetails testDetails = testDetailsRepository.findById(testId);
+            Desktop.getDesktop().open(new File(testDetails.getPathPDF()));
```
Comment (schneijan) [inline src/main/java/org/citopt/connde/service/testing/TestEngine.java:427]: Seems like this opens the PDF file on the desktop of the server? But from the perspective of the user this only works if the user is running the MBP on his machine himself, doesn't it? I think this should be removed or replaced by a download that is offered to the user.

File: src/main/java/org/citopt/connde/service/testing/TestReport.java [hunk 2]
```diff
@@ -0,0 +1,627 @@
...
+
+        // Create a new pdf, which is named with the ID of the specific test
+        File tempFile = new File(testId + ".pdf");
+        if (tempFile.exists() && tempFile.isFile()) {
+            tempFile.delete();
+        }
+        File testReport = new File(testId + ".pdf");
+        String path = testReport.getAbsolutePath();
+
+        FileOutputStream pdfFileout = new FileOutputStream(testReport);
```
Comment (schneijan) [inline src/main/java/org/citopt/connde/service/testing/TestReport.java:69]: Would it maybe be possible to store the file somewhere else, e.g. in a database or to generate the report on demand without creating a file? Storing files on disk makes the server stateful and breaks the scalability of the application.
In addition, it looks like this file is never deleted after creation, leading to a full disk at some point in time.

**Comment by hirmerpl** [comment]
The issue is with JFrame, java headless JDK as it comes with Ubuntu does not support JFrame or the other AWT libraries. Maybe they can be replaced with other libraries? Alternatively, we would need another JDK.

**Comment by hirmerpl** [comment]
Download PDF does not work for me currently.

**Pull request merged**